Add Store Widgets target and Fastlane configuration

mokagio opened the pull request:
### Description

Add a new widgets extension target and configures the Fastlane automation to code sign it. Internal ref pdnsEh-yV-p2.

The diff is unfortunately long, as a byproduct of adding the target. However, all the code in `StoreWidgets` was generated by Xcode and I didn't modify it's functionality. So, the only files to review in this PR are `Fastfile` and the `project.pbxproj`. Conveniently, they are both at the very bottom of the diff.

### Testing instructions

If CI succeeds in building the Enterprise build, then we're good.

I also created a hack PR to test that CI can build the App Store build [here](https://github.com/woocommerce/woocommerce-ios/pull/7624)—It passed.


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary. – _The addition of the widgets target is a major one, but this PR has not feature change from that point of view._

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

mokagio committed:
Add "Widget Extension" target via Xcode UI

mokagio committed:
Use `storewidgets` instead of `StoreWidgets` in the bundle id

This is the value requested by the dev team and it was not possible to
set it via the Xcode wizard at creation time.

mokagio committed:
Extract App Store bundle identifier in constant in `Fastfile`

We'll use it next to build the widget extension bundle id in a
parametric way.

mokagio committed:
Extract Alpha/Enterprise bundle identifier in constant in `Fastfile`

We'll use it next to build the widget extension bundle id in a
parametric way.

mokagio committed:
Make `match` run in `readonly` mode for both ASC and Enterprise

Always run `readonly` by default, just in case.

mokagio committed:
Add widgets extension bundle identifier to bundle arrays in `Fastfile`

mokagio committed:
Update Fastlane to latest version, 2.209.1

Ecarrion committed:
Merge branch 'trunk' into mokagio/add-widgets-target

Ecarrion committed:
Merge branch 'trunk' into mokagio/add-widgets-target

mokagio committed:
Merge remote-tracking branch 'origin/trunk' into mokagio/add-widgets-target

mokagio commented on WooCommerce/StoreWidgets/StoreWidgets.swift:
```diff
@@ -0,0 +1,70 @@
+    let configuration: ConfigurationIntent
+}
+
+struct StoreWidgetsEntryView : View {
```
Apple having inconsistent style _in the same template file!_

mokagio committed:
Remove space in between `struct` name and `:` in `StoreWidgets.swift`

mokagio committed:
Remove custom entitlements from `match` App Store configuration

mokagio commented on fastlane/Fastfile:
```diff
@@ -932,18 +943,24 @@ platform :ios do
     match(
       type: 'enterprise',
       team_id: get_required_env('INT_EXPORT_TEAM_ID'),
-      readonly: true,
-      app_identifier: ALPHA_BUNDLE_IDENTIFIERS
+      app_identifier: ALPHA_BUNDLE_IDENTIFIERS,
+      readonly: true
```
I moved this around to keep it consistent with the one below. 

I moved `readonly: true` at the bottom below to have a parameter less likely to change at the end of the list and avoid awkward diffs where the only change is adding or removing a `,` to the parameter that used to be last but no longer is.

mokagio commented on fastlane/Fastfile:
```diff
@@ -932,18 +943,24 @@ platform :ios do
-      readonly: false,
       app_identifier: MAIN_BUNDLE_IDENTIFIERS,
-      template_name: 'Apple Pay Pass Suppression'
+      # For a time, we used to have this parameter specifying a entitlements.
+      # In August 2022, `match` started being unable to generate new
+      # provisioning profiles and the only solution I found was to remove it.
+      #
+      # Leaving it here for future referece as something to try if `match`
+      # gives us trouble again.
+      # template_name: 'Apple Pay Pass Suppression'
```
I'm not sure whether more detail is needed here? I'm not keen in adding links to internal documents (even if in shorthand form) in the code. There should be detail here to search for the information internally if necessary. What do you think?

AliSoftware replied:
```suggestion
      # For a time, we used to have this parameter specifying a entitlements.
      # In August 2022, `match` started being unable to generate new
      # provisioning profiles and the only solution I found was to remove it.
      #
      # Leaving it here for future reference as something to try if `match`
      # gives us trouble again.
      # template_name: 'Apple Pay Pass Suppression'
```

AliSoftware replied:
At first I misread (multiple times, even… ☕ ) this as "`match` started being **able** to […]" and that sentence got me so confused, because I was like "so wait, match got better with a new feature, and you had to find a way to… fix it?". Stumbled me for two good minutes… until I was about to make a GitHub suggestion to reword this and finally notice that you wrote "unable", not "able".
I think the "starting being" wording led me astray, as I'm not used in hearing "start to be unable" and would expect a wording like "stop being able" instead.

Also I'm not sure if that wording is supposed to mean that this is a bug in `match` (it started to have a bug which made it fail to create new profiles) or a feature (they purposely removed that ability because it wasn't working as expected or whatnot, and thus removed the `template_name` parameter as well).

Your subsequent sentence suggesting that we might re-introduce the `template_name` parameter again in the future lead me to finally understand that this might probably be a bug and not a conscious decision by fastlane, but the "to try if match gives us trouble again" added to the confusion 😅  As in: if the bug was when match tried to create new profiles because of this parameter, how come match would… give us trouble again without it? If anything, I feel that the reason why we'd want to add the parameter back is if match fixed the profile creation bug and we wanted to give it back a try, no?

mokagio replied:
The truth is... I don't know if this is a bug in `match`, a bug on Apple's side, or something that never worked, was worked around in a different way, and we only found out about it now because we decided to generate new profiles 🙃 

It could be the latter, because the [automation](https://github.com/Automattic/match-regenerator/blob/edef2ca14b0f62ccfab0f6426711daf296d56b98/fastlane/Fastfile) I cobbled together in the past to regenerate the provisioning profiles across all app never had `template_name` and worked back then. I had forgotten about that code and only though of it in writing this.

I think I'm just going to remove the comment. Git remembers.

AliSoftware replied:
👍 on removing the comment. git blame ftw 😛

mokagio commented:
Note that the [Installable Builds failure](https://buildkite.com/automattic/woocommerce-ios/builds/7537#0182ec81-951b-4c8b-977d-0c7845cefde2) is expected at this time because of unrelated issues accessing the Enterprise account.

<img width="975" alt="image" src="https://user-images.githubusercontent.com/1218433/187335417-5552a886-9c8c-45db-8817-0852617ae144.png">

mokagio committed:
Temporarily remove widgets bundle id from Enterprise `match` call

Such a setup is no good long term but we're obviously working to sort it
out. In the short term, not including this bundle id might work, or it
might break the build later on when code sign happens but can't find the
provisioning profile. We'll soon find out...

mokagio commented on fastlane/Fastfile:
```diff
@@ -36,7 +36,14 @@ ALPHA_VERSION_BUNDLE_IDENTIFIER = 'com.automattic.alpha.woocommerce'
+  # out. In the short term, not including this bundle id might work, or it
+  # might break the build later on when code sign happens but can't find the
+  # provisioning profile. We'll soon find out...
+  # "#{ALPHA_VERSION_BUNDLE_IDENTIFIER}.storewidgets"
```
As expected, the build failed during code signing:

<img width="1170" alt="image" src="https://user-images.githubusercontent.com/1218433/187376623-a622560e-0a4f-4f00-8a02-25c2ab8afcfd.png">

```
error: StoreWidgetsExtension has conflicting provisioning settings. StoreWidgetsExtension is automatically signed for development, 
but a conflicting code signing identity iPhone Distribution has been manually specified.
Set the code signing identity value to "Apple Development" in the build settings editor, or switch to manual signing in the Signing & Capabilities editor.
(in target 'StoreWidgetsExtension' from project 'WooCommerce')
```

mokagio committed:
Merge remote-tracking branch 'origin/trunk' into mokagio/add-widgets-target

mokagio committed:
Restore widgets bundle id in Enterprise `match` call

This reverts commit 30d3b54df89f80563e348649c2ed34075710545c.

The issue that was preventing us from accessing the Enterprise account
to provision the Application Identifier for the new extension has been
solved.

I run the automation to update the Enterprise app code signing
separately. CI should now be able to archive the app.

mokagio committed:
Update StoreWidgets configuration to work with `match`

mokagio committed:
Update `ALWAYS_EMBED_SWIFT_STANDARD_LIBRARIES` setting in WooCommerce

Xcode set it to `YES` when creating the Widgets Extension target in
bfebd4c48cb473619bac8a9460cc9308aec6657c, but we need to account for our
CocoaPods setup which already takes care of it.

To verify this, run `bundle exec pod install` from this commit and in
the previous one. You'll notice the previous logs the following warning:

```
[!] The `WooCommerce [Debug]` target overrides the `ALWAYS_EMBED_SWIFT_STANDARD_LIBRARIES` build setting defined in `../Pods/Target Support Files/Pods-WooCommerce/Pods-WooCommerce.debug.xcconfig'. This can lead to problems with the CocoaPods installation
    - Use the `$(inherited)` flag, or
    - Remove the build settings from the target.

[!] The `WooCommerce [Release]` target overrides the `ALWAYS_EMBED_SWIFT_STANDARD_LIBRARIES` build setting defined in `../Pods/Target Support Files/Pods-WooCommerce/Pods-WooCommerce.release.xcconfig'. This can lead to problems with the CocoaPods installation
    - Use the `$(inherited)` flag, or
    - Remove the build settings from the target.

[!] The `WooCommerce [Release-Alpha]` target overrides the `ALWAYS_EMBED_SWIFT_STANDARD_LIBRARIES` build setting defined in `../Pods/Target Support Files/Pods-WooCommerce/Pods-WooCommerce.release-alpha.xcconfig'. This can lead to problems with the CocoaPods installation
    - Use the `$(inherited)` flag, or
    - Remove the build settings from the target.
```

Also notice that the final computed value for the build setting in the
Xcode UI remains "Yes".

mokagio commented on fastlane/Fastfile:
```diff
@@ -27,8 +27,19 @@ SCREENSHOT_DEVICES = [
 
-MAIN_BUNDLE_IDENTIFIERS = %w[com.automattic.woocommerce].freeze        # Registered in our main account, for development and AppStore
-ALPHA_BUNDLE_IDENTIFIERS = %w[com.automattic.alpha.woocommerce].freeze # Registered in our Enterprise account, for App Center / Installable Builds
+APP_STORE_VERSION_BUNDLE_IDENTIFIER = 'com.automattic.woocommerce'
+# Registered in our main account, for development and App Store
+MAIN_BUNDLE_IDENTIFIERS = [
+  APP_STORE_VERSION_BUNDLE_IDENTIFIER,
+  "#{APP_STORE_VERSION_BUNDLE_IDENTIFIER}.storewidgets"
+].freeze
+
+ALPHA_VERSION_BUNDLE_IDENTIFIER = 'com.automattic.alpha.woocommerce'
+# Registered in our Enterprise account, for App Center / Installable Builds
+ALPHA_BUNDLE_IDENTIFIERS = [
+  ALPHA_VERSION_BUNDLE_IDENTIFIER,
+  "#{ALPHA_VERSION_BUNDLE_IDENTIFIER}.storewidgets"
+].freeze
```
In WordPress iOS, we have a [fancier, DRYer setup](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/fastlane/Fastfile#L30-L43) for the bundle identifiers. I didn't port it across because I found it less readable than this one. It makes sense in WordPress where we have 9 extensions, but I think this works fine here.

Happy to update in a dedicated PR if you feel it would be good for future proofing.

Ecarrion reviewed (approved):
I can't comment much about the `fastfile` but the setup is working correctly for my initial widget UI. https://github.com/woocommerce/woocommerce-ios/pull/7584

mokagio committed:
Remove comment regarding `template_name: 'Apple Pay Pass Suppression'`

It was partly misleading. See conversation at
https://github.com/woocommerce/woocommerce-ios/pull/7558/files#r957944888

mokagio commented:
@Ecarrion thank you for looking into this. I think if we merge it as is, the dummy widget will be available to users. That would be confusing to say the least. 

Might it be best to merge this only when the widget is usable enough?

Ecarrion commented:
Hey @mokagio , I think we can merge this as is, as we already have an open PR that hides the widget under a local feature flag.

Ecarrion commented:
@mokagio I'm going to merge this so we can advance with the development on trunk! Thanks a lot again! 🙇

AliSoftware commented:
@mokagio seems that `trunk` broke after this PR landed… At least some people in Slack (@jaclync, @joshheald), are mentioning they can't build the app on device in trunk due to provisioning profile for the widget target is missing. (ref: p1662368046975199-slack-CC7L49W13). Any idea?
